### PR TITLE
SequenceFusion transformation

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
 - *The transformation fuses a chain of GRU/AUGRU/RNN/LSTM Cells to GRU/AUGRU/RNN/LSTM  Sequence operation.*

### Tickets:
 - *87890, 86816*

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        // GRUCell -> GRUCell (the last cell) -> OtherNode
+        // GRUCell (hidden_size = 128) -> GRUCell (hs = 128, the last) -> GRUCell (hs = 64)
+        for (const auto& target : cell->get_output_target_inputs(0)) {
+            auto cell_1 = dynamic_pointer_cast<RNNCellBase>(target.get_node()->shared_from_this());
+            if (cell_1 && is_equal_cells(cell_1, current_cell)) {
+                return false;
+            }
+        }
+
+        int cells_cnt = 0;
```
Comment (praasz) [inline src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:279]: I think change it to 
```suggestion
        int cells_cnt;
```
As it should be set in the function below. If not happens e.g. static analysis tools should complain that working on uninitialized variable which can be suggestion of programming error set it will mask all issues.

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+#include "ngraph_ops/augru_sequence.hpp"
+#include "node_register.hpp"
+
+#include "openvino/core/rt_info.hpp"
+#include "openvino/opsets/opset9.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+using namespace std;
+using namespace ov;
+using namespace ov::opset9;-
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: already fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+shared_ptr<RNNCellBase> find_cell_chain(ov::pass::NodeRegister& cp_from,
+                                        ov::pass::NodeRegister& cp_to,
+                                        const shared_ptr<RNNCellBase>& current_cell,
+                                        OutputVector& x_to_concat,
+                                        OutputVector& attention_to_concat,
+                                        map<int, set<ov::Input<Node>>>& h_inputs_to_redirect,
+                                        map<int, set<ov::Input<Node>>>& c_inputs_to_redirect,
+                                        int& cells_cnt,
+                                        const shared_ptr<Node>& axis_1) {
+    cells_cnt++;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:85]: this is the count of Cells in the chain
e.g. (OtherNode->Cell->Cell->Cell->Cell->OtherNode), cells_cnt = 4
it's better to initialize it to 1, I will change it and will add short description for this function

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    auto lW = cell_1->input_value(idx_W).get_node_shared_ptr();
+    auto lR = cell_1->input_value(idx_R).get_node_shared_ptr();
+    auto lB = cell_1->input_value(idx_B).get_node_shared_ptr();
+    auto rW = cell_2->input_value(idx_W).get_node_shared_ptr();
+    auto rR = cell_2->input_value(idx_R).get_node_shared_ptr();
+    auto rB = cell_2->input_value(idx_B).get_node_shared_ptr();
+    bool is_equal = lW.get() == rW.get() || is_equal_consts(lW, rW);
+    is_equal &= lR.get() == rR.get() || is_equal_consts(lR, rR);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:56]: I believe the result is always correct in C++ but yes, you are right about the evaluation of the expression, it always executed.
I will change this

File: src/common/transformations/include/transformations/common_optimizations/sequence_fusion.hpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+}  // namespace pass
+}  // namespace ov
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief SequenceFusion transformation replaces a chain of GRUCell
+ * operations with single GRUSequence op.
+ *
+ * Supported cells: GRUCell, LSTMCell, RNNCell, AUGRUCell
+ * Limitations: the source of W,R,B inputs must be the same or
```
Comment (ceciliapeng2011) [inline src/common/transformations/include/transformations/common_optimizations/sequence_fusion.hpp:26]: I don't think it is the limitation. It is the prerequisite instead.

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp [hunk 6]
```diff
@@ -178,6 +179,7 @@ bool ngraph::pass::MOCTransformations::run_on_model(const std::shared_ptr<ngraph
     common_fusions->add_matcher<ngraph::pass::PReluFusion>();
     common_fusions->add_matcher<ngraph::pass::DepthToSpaceFusion>();
     common_fusions->add_matcher<ngraph::pass::ShuffleChannelsFusion>(!m_use_shapes);
+    common_fusions->add_matcher<ov::pass::SequenceFusion>();
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:182]: performance measurements needed

File: src/core/src/graph_util.cpp [hunk 7]
```diff
@@ -749,51 +749,65 @@ bool ngraph::check_for_cycles(const ngraph::Function* func, ngraph::NodeVector&
...
 }
 
+bool has_result_consumers(const Output<Node>& port) {
+    const auto& consumers = port.get_target_inputs();
+    return std::any_of(consumers.cbegin(), consumers.cend(), [](const Input<Node>& consumer) {
+        return ov::is_type<op::v0::Result>(consumer.get_node());
+    });
+};
+
+bool ov::replace_outputs_update_names(OutputVector outputs, const OutputVector& replacements) {
```
Comment (itikhono) [inline src/core/src/graph_util.cpp:759]: Changes in this file will be moved to the separate PR. 
This is not a part of the transformation.

File: src/common/transformations/include/transformations/common_optimizations/sequence_fusion.hpp (hunk 5 above)
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/sequence_fusion.hpp]: fixed

File: src/tests/functional/inference_engine/transformations/common_optimizations/sequence_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    int64_t cell_cnt;
+};
+
+class SequenceFusionTest
+        : public testing::WithParamInterface<SequenceFusionParams>,
+          public TransformationTestsF {
+};
+
+TEST_P(SequenceFusionTest, SequencePattern) {
+    auto p = GetParam();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/sequence_fusion.cpp]: fixed

File: src/tests/functional/inference_engine/transformations/common_optimizations/sequence_fusion.cpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    if (p.cell_cnt > 1) {
+        function_ref = gen_reference(p.rnn_type, p.batch, p.hidden_size, p.input_size, p.cell_cnt);
+    }
+    comparator.enable(FunctionsComparator::CmpValues::ACCURACY);
+    comparator.enable(FunctionsComparator::CmpValues::CONST_VALUES);
+    comparator.enable(FunctionsComparator::CmpValues::ATTRIBUTES);
+}
+
+static std::vector<SequenceFusionParams> params = {
+        SequenceFusionParams{RNN_TYPE::GRU, 2, 128, 32, 10},
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/sequence_fusion.cpp:197]: const added

File: src/tests/functional/inference_engine/transformations/common_optimizations/sequence_fusion.cpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+        SequenceFusionParams{RNN_TYPE::RNN, 2, 128, 32, 10},
+        SequenceFusionParams{RNN_TYPE::AUGRU, 2, 128, 32, 10},
+};
+
+INSTANTIATE_TEST_SUITE_P(SequenceFusionTest, SequenceFusionTest, ::testing::ValuesIn(params));
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/tests/functional/inference_engine/transformations/common_optimizations/sequence_fusion.cpp]: fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+        if (!res) {
+            return false;
+        }
+        copy_runtime_info(copy_from.get(), copy_to.get());
+        return true;
+    };
+
+    auto m = make_shared<Matcher>(cell, matcher_name);
+    this->register_matcher(m, callback);
+}
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/sequence_fusion.hpp"
+
+#include <memory>
+
+#include "itt.hpp"
+#include "ngraph_ops/augru_cell.hpp"
+#include "ngraph_ops/augru_sequence.hpp"
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:11]: I will create a ticket

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: fixed

[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: still need to add the description

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+bool check_WRB(const shared_ptr<RNNCellBase>& cell_1, const shared_ptr<RNNCellBase>& cell_2) {
+    int64_t idx_W = 2, idx_R = 3, idx_B = 4;
+    auto lstm_cell_1 = dynamic_pointer_cast<LSTMCell>(cell_1);
+    auto lstm_cell_2 = dynamic_pointer_cast<LSTMCell>(cell_2);
+
+    // 2nd input is Cell State
+    if (lstm_cell_1 && lstm_cell_2) {
+        idx_B++;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: fixed

File: src/common/transformations/include/transformations/common_optimizations/sequence_fusion.hpp [hunk 14]
```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <openvino/pass/graph_rewrite.hpp>
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/sequence_fusion.hpp:8]: fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+bool is_equal_cells(const shared_ptr<RNNCellBase>& cell_1, const shared_ptr<RNNCellBase>& cell_2) {
+    bool is_equal =
+        cell_1->get_type_name() == cell_2->get_type_name() && cell_1->get_hidden_size() == cell_2->get_hidden_size() &&
+        cell_1->get_activations() == cell_2->get_activations() &&
+        cell_1->get_activations_alpha() == cell_2->get_activations_alpha() &&
+        cell_1->get_activations_beta() == cell_2->get_activations_beta() && cell_1->get_clip() == cell_2->get_clip();
+    is_equal &= check_WRB(cell_1, cell_2);
+    auto gru_cell_1 = dynamic_pointer_cast<GRUCell>(cell_1);
+    auto gru_cell_2 = dynamic_pointer_cast<GRUCell>(cell_2);
+    if (gru_cell_1 && gru_cell_2) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:66]: fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+        is_equal &= gru_cell_1->get_linear_before_reset() == gru_cell_2->get_linear_before_reset();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: fixed

File: src/core/src/graph_util.cpp [hunk 17]
```diff
@@ -32,6 +32,17 @@ using namespace std;
 
 NGRAPH_SUPPRESS_DEPRECATED_START
 
+namespace {
+
+bool has_result_consumers(const ov::Output<ov::Node>& port) {
+    const auto& consumers = port.get_target_inputs();
+    return std::any_of(consumers.cbegin(), consumers.cend(), [](const ov::Input<ov::Node>& consumer) {
+        return ov::is_type<op::v0::Result>(consumer.get_node());
```
Comment (itikhono) [inline src/core/src/graph_util.cpp:40]:
```suggestion
        return ov::is_type<ov::op::v0::Result>(consumer.get_node());
```

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+                                        const shared_ptr<RNNCellBase>& current_cell,
+                                        OutputVector& x_to_concat,
+                                        OutputVector& attention_to_concat,
+                                        map<int, set<ov::Input<Node>>>& h_inputs_to_redirect,
+                                        map<int, set<ov::Input<Node>>>& c_inputs_to_redirect,
+                                        int& cells_cnt,
+                                        const shared_ptr<Node>& axis_1) {
+    cells_cnt++;
+
+    shared_ptr<RNNCellBase> current = current_cell;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:88]: we use cell_cnt in the other function (create_sequence), it can't be internal variable here

File: src/core/src/graph_util.cpp [hunk 19]
```diff
@@ -749,51 +749,65 @@ bool ngraph::check_for_cycles(const ngraph::Function* func, ngraph::NodeVector&
...
+        return false;
+    }
+
+    for (int64_t idx = 0; idx < outputs.size(); ++idx) {
+        auto& output = outputs[idx];
+        const auto& replacement = replacements[idx];
+        bool preserve_legacy_output_name = false;
+        if (has_result_consumers(output)) {
+            preserve_legacy_output_name = true;
+            if (is_type<ngraph::op::Parameter>(replacement.get_node()) || has_result_consumers(replacement)) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/core/src/graph_util.cpp:774]: the changes have been reverted

File: src/core/src/graph_util.cpp [hunk 20]
```diff
@@ -749,51 +749,65 @@ bool ngraph::check_for_cycles(const ngraph::Function* func, ngraph::NodeVector&
...
+
+bool ov::replace_outputs_update_names(OutputVector outputs, const OutputVector& replacements) {
+    if (outputs.size() != replacements.size()) {
+        return false;
+    }
+
+    for (int64_t idx = 0; idx < outputs.size(); ++idx) {
+        auto& output = outputs[idx];
+        const auto& replacement = replacements[idx];
+        bool preserve_legacy_output_name = false;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/core/src/graph_util.cpp:771]: the changes have been reverted

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 21]
```diff
@@ -0,0 +1,325 @@
...
+            auto cell_1 = dynamic_pointer_cast<RNNCellBase>(target.get_node()->shared_from_this());
+            if (cell_1 && is_equal_cells(cell_1, current_cell)) {
+                return false;
+            }
+        }
+
+        int cells_cnt = 0;
+        OutputVector x_to_concat;
+        OutputVector attention_to_concat;
+        map<int, set<ov::Input<Node>>> h_inputs_to_redirect;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: using namespace ov; causes the conflicts in other transformations

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:11]: ticket number 92393

**Review by praasz — APPROVED** [review]

**Review by jane-intel — APPROVED** [review]
Looks legit to me. We have LSTMCell from 1 opset and from 4 -- how will we deal with LSTM-1?

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+
+    shared_ptr<RNNCellBase> current = current_cell;
+    while (true) {
+        cp_from.add(current);
+        // check the source node of HiddenState input
+        auto prev = current->input_value(1).get_node_shared_ptr();
+        if (auto prev_cell = dynamic_pointer_cast<RNNCellBase>(prev)) {
+            if (!is_equal_cells(prev_cell, current)) {
+                break;
+            }
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:94]: (Just a corner case)
What if the "not equal" cell will be at the end of 20 equal cells? 
The code from `else` block won't be reached. Is it ok?

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp [hunk 23]
```diff
@@ -0,0 +1,324 @@
...
+            cells_cnt++;
+        } else {
+            auto in_X = current->input(0);
+            x_to_concat.push_back(cp_to.make<Unsqueeze>(in_X.get_source_output(), axis_1));
+            if (auto augru = dynamic_pointer_cast<ov::op::internal::AUGRUCell>(current)) {
+                attention_to_concat.push_back(cp_to.make<Unsqueeze>(augru->input_value(5), axis_1));
+            }
+            h_outputs_to_redirect[cells_cnt] = current->output(0);
+            break;
+        }
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp:127]: Looks like this piece of code in the `else` block could be moved out of the `if/else` as it is repeated twice.
https://github.com/openvinotoolkit/openvino/blob/20508c9aeb741bb767f25b1566273fa7dac3bfbd/src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp#L96-L98 https://github.com/openvinotoolkit/openvino/blob/20508c9aeb741bb767f25b1566273fa7dac3bfbd/src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp#L113-L115

**Review by mitruska — APPROVED** [review]

**Review by ceciliapeng2011 — APPROVED** [review]
LGTM.

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp (hunk 22 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: not ok :) thanks
fixed

File: src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp (hunk 23 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/sequence_fusion.cpp]: done

**Review by ilyachur — APPROVED** [review]
